### ISSUE TITLE
can't get timestamp from JSON log with promtail

### ISSUE THREAD

**Issue by ufth** [grafana/loki#3694]
I've already spend almost a day trying to get a proper timestamp from
nginx logs in JSON format to be sure I can see it in Grafana - e.g.
not the exact time when this log line was shipped to loki, but an
exact time when this log line was originally created. I was able to
achieve this goal with usual syslog log format quickly, but I have no
luck with JSON and nginx.

Log line example:

```
 head -n 2 json_access.log
{"msec": "1619676109.299", "connection": "95359955",
"connection_requests": "1", "pid": "25755", "request_id":
"cf0a6bf1e469860b94fa76f48fd0236d", "request_length": "457",
"remote_addr": "93.158.161.42", "remote_user": "", "remote_port": "",
"time_local": "29/Apr/2021:09:01:49 +0300", "time_iso8601":
"2021-04-29T09:01:49+03:00", "request": "GET /ru/yandex HTTP/1.1",
"request_uri": "/ru/yandex", "args": "", "status": "200",
"body_bytes_sent": "54964", "bytes_sent": "55720", "http_referer": "",
"http_user_agent": "Mozilla/5.0 (compatible; YandexBot/3.0;
+http://yandex.com/bots)", "http_x_forwarded_for": "93.158.161.42",
"http_host": "example.com", "server_name": "example.com",
"request_time": "0.111", "upstream": "127.0.0.1:9001",
"upstream_connect_time": "0.000", "upstream_header_time": "0.108",
"upstream_response_time": "0.112", "upstream_response_length":
"159471", "upstream_cache_status": "EXPIRED", "ssl_protocol": "",
"ssl_cipher": "", "scheme": "http", "request_method": "GET",
"server_protocol": "HTTP/1.1", "pipe": ".", "gzip_ratio": "2.90",
"http_cf_ray": "64765160fbf27a7d-DME", "request_completion": "OK"}
{"msec": "1619676114.113", "connection": "95359900",
"connection_requests": "1", "pid": "25755", "request_id":
"56ec9ddb6a2fda08c2ea7283c8ddc504", "request_length": "677",
"remote_addr": "66.249.75.164", "remote_user": "", "remote_port": "",
"time_local": "29/Apr/2021:09:01:54 +0300", "time_iso8601":
"2021-04-29T09:01:54+03:00", "request": "GET /someurl", "args":
"Name=bill", "status": "200", "body_bytes_sent": "53794",
"bytes_sent": "54028", "http_referer": "", "http_user_agent":
"Mozilla/5.0 (compatible; Googlebot/2.1;
+http://www.google.com/bot.html)", "http_x_forwarded_for":
"66.249.75.164", "http_host": "example.com", "server_name":
"example.com", "request_time": "22.727", "upstream": "127.0.0.1:9002",
"upstream_connect_time": "0.000", "upstream_header_time": "22.623",
"upstream_response_time": "22.727", "upstream_response_length":
"647782", "upstream_cache_status": "", "ssl_protocol": "",
"ssl_cipher": "", "scheme": "http", "request_method": "GET",
"server_protocol": "HTTP/1.1", "pipe": ".", "gzip_ratio": "12.04",
"http_cf_ray": "647650f2ee470389-ORD", "request_completion": "OK"}
```

**To Reproduce**
1. Started Loki 2.2.1
2. Started Promtail 2.2.1 to tail /var/log/nginx/json_access.log
3. Query logs in Grafana

**Expected behavior**
Timestamp in Grafana should be taken from the log line imported

**Actual behavior**
Timestamp in Grafana shows time when this line was received by loki

**Environment:**
 - bare-metal, laptop, vps
 - promtail machine: Debian GNU/Linux 9 \n \l ; Linux xxx 4.9.0-11-amd64 #1 SMP Debian 4.9.189-3+deb9u2 (2019-11-11) x86_64 GNU/Linux
 - loki machine: Debian GNU/Linux 10 \n \l ; Linux yyy 4.19.0-16-cloud-amd64 #1 SMP Debian 4.19.181-1 (2021-03-19) x86_64 GNU/Linux
 - Deployment tool: none, manually downloaded and unzipped binaries

**Promtail config**

```
cat promtail-local-config.yaml
server:
  http_listen_port: 9080
  grpc_listen_port: 0

positions:
  filename: /data/loki/positions.yaml

clients:
  - url: http://localhost:3100/loki/api/v1/push

scrape_configs:
 - job_name: system
   pipeline_stages:
   - match:
       selector: '{job="nginx_acess_log"}'
       stages:
       - json:
           expressions:
             timestampp: time_iso8601
             #timestampp: msec
       - timestamp:
           source:  timestampp
           format: '2006-06-02T03:04:05-07:00'
           #format: 'Unix'
           #format: 'RFC3339'
   static_configs:
   - targets:
      - localhost
     labels:
      job: nginx_access_log
      host: yourhost
      __path__: /var/log/nginx/json_access.log


./loki-linux-amd64 --version
loki, version 2.2.1 (branch: HEAD, revision: babea82e)
  build user:       root@e2d295b84e26
  build date:       2021-04-06T00:52:41Z
  go version:       go1.15.3
  platform:         linux/amd64

./promtail-linux-amd64 --version
promtail, version 2.2.1 (branch: HEAD, revision: babea82e)
  build user:       root@e2d295b84e26
  build date:       2021-04-06T00:52:41Z
  go version:       go1.15.3
  platform:         linux/amd64
```

I tired time_iso8601 and msec fields and different formats (pre-defined or manually defined) without any
result. What am I doing wrong? Is there a convenient way to debug this
exact thing?

**Comment by Mario-Hofstaetter** [comment]
I have not actually tried parsing your log, yet, but looking at your examples and the format you tried:

```json
"time_iso8601":"2021-04-29T09:01:49+03:00"
"time_iso8601":"2021-04-29T09:01:54+03:00"
```

```yaml
       - timestamp:
           source:  timestampp
           format: '2006-06-02T03:04:05-07:00' # ❌ 
```

According to <https://golang.org/src/time/format.go>:
- Month as `06` is incorrect ❌ , it should be `01` (`stdZeroMonth`)
- Hour as `03` is 12h format ❌ , I'm sure nginx uses 24h format, so it should be `15` (`stdHour`)

So could you try this out:
```yaml
       - timestamp:
           source:  timestampp
           format: '2006-01-02T15:04:05-07:00'
```

**Comment by ufth** [comment]
Thank you very much for your quick response, @Mario-Hofstaetter ! I tried the new format suggested but still had no luck - see the screenshot below:

![promtail-new-timestamp-format-suggested](https://user-images.githubusercontent.com/8803171/117408751-0fdb3700-af19-11eb-9774-aa7a2353ff5f.png)

**Comment by ufth** [comment]
So, this is not a bug. 

I was able to solve the issue playing with `-dry-run` and `-log.level=debug` parameters for promtail using stdin as an input and a correct date format kindly suggested by @Mario-Hofstaetter. My initial YAML configuration file was completely incorrect and I was able to detect that timstamp block is never triggered, so I can use anything as "format" and there is no error. So I created  whole config file from the scratch using my valid "syslog" config as an example.

![01](https://user-images.githubusercontent.com/8803171/117871081-a905a700-b2a5-11eb-9776-69a8a8d71f36.png) 

Please note, that you should see your timestamp from the logs printed in blue if everything is correct. If something went wrong, this blue date will be your current date or you can see an error like this (let's try to apply our format to "msec" field):
`level=debug ts=2021-05-11T19:14:43.380825475Z caller=timestamp.go:196 component=pipeline msg="failed to parse time" err="parsing time \"1619676114.113\" as \"2006-01-02T15:04:05-07:00\": cannot parse \"676114.113\" as \"-\"" format=2006-01-02T15:04:05-07:00 value=1619676114.113
`

```
 cat json_access.log.small | ./promtail --stdin -config.file promtail-local-config.yaml -log.level=debug -dry-run
Clients configured:
----------------------
url: http://localhost:3100/loki/api/v1/push
batchwait: 1s
batchsize: 1048576
backoff_config:
  min_period: 500ms
  max_period: 5m0s
  max_retries: 10
timeout: 10s
tenant_id: ""

level=debug ts=2021-05-11T19:04:26.03229948Z caller=manager.go:58 msg="configured to read from stdin"
level=info ts=2021-05-11T19:04:26.033040649Z caller=server.go:229 http=[::]:9080 grpc=[::]:38187 msg="server listening on addresses"
level=info ts=2021-05-11T19:04:26.033309926Z caller=main.go:112 msg="Starting Promtail" version="(version=2.2.1, branch=HEAD, revision=babea82e)"
2021-04-29T09:01:49     {__path__="/var/log/nginx/json_access.log", host="yourhost", job="nginx_access_log"}    {"msec": "1619676109.299", "connection": "95359955", "connection_requests": "1", "pid": "25755", "request_id": "cf0a6bf1e469860b94fa76f48fd0236d", "request_length": "457", "remote_addr": "93.158.161.42", "remote_user": "", "remote_port": "", "time_local": "29/Apr/2021:09:01:49 +0300", "time_iso8601": "2021-04-29T09:01:49+03:00", "request": "GET /ru/yandex HTTP/1.1", "request_uri": "/ru/yandex", "args": "", "status": "200", "body_bytes_sent": "54964", "bytes_sent": "55720", "http_referer": "", "http_user_agent": "Mozilla/5.0 (compatible; YandexBot/3.0; +http://yandex.com/bots)", "http_x_forwarded_for": "93.158.161.42", "http_host": "example.com", "server_name": "example.com", "request_time": "0.111", "upstream": "127.0.0.1:9001", "upstream_connect_time": "0.000", "upstream_header_time": "0.108", "upstream_response_time": "0.112", "upstream_response_length": "159471", "upstream_cache_status": "EXPIRED", "ssl_protocol": "", "ssl_cipher": "", "scheme": "http", "request_method": "GET", "server_protocol": "HTTP/1.1", "pipe": ".", "gzip_ratio": "2.90", "http_cf_ray": "64765160fbf27a7d-DME", "request_completion": "OK"}
2021-04-29T09:01:54     {__path__="/var/log/nginx/json_access.log", host="yourhost", job="nginx_access_log"}    {"msec": "1619676114.113", "connection": "95359900", "connection_requests": "1", "pid": "25755", "request_id": "56ec9ddb6a2fda08c2ea7283c8ddc504", "request_length": "677", "remote_addr": "66.249.75.164", "remote_user": "", "remote_port": "", "time_local": "29/Apr/2021:09:01:54 +0300", "time_iso8601": "2021-04-29T09:01:54+03:00", "request": "GET /someurl", "args": "Name=bill", "status": "200", "body_bytes_sent": "53794", "bytes_sent": "54028", "http_referer": "", "http_user_agent": "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)", "http_x_forwarded_for": "66.249.75.164", "http_host": "example.com", "server_name": "example.com", "request_time": "22.727", "upstream": "127.0.0.1:9002", "upstream_connect_time": "0.000", "upstream_header_time": "22.623", "upstream_response_time": "22.727", "upstream_response_length": "647782", "upstream_cache_status": "", "ssl_protocol": "", "ssl_cipher": "", "scheme": "http", "request_method": "GET", "server_protocol": "HTTP/1.1", "pipe": ".", "gzip_ratio": "12.04", "http_cf_ray": "647650f2ee470389-ORD", "request_completion": "OK"}
```

Correct configuration file:

```
cat promtail-local-config.yaml
server:
  http_listen_port: 9080
  grpc_listen_port: 0

positions:
  filename: /data/loki/positions.yaml

clients:
  - url: http://localhost:3100/loki/api/v1/push

scrape_configs:
 - job_name: system
   static_configs:
   - targets:
      - localhost
     labels:
      job: nginx_access_log
      host: yourhost
      __path__: /var/log/nginx/json_access.log
   pipeline_stages:
   - match:
       selector: '{job="nginx_access_log"}'
       stages:
       - json:
           expressions:
             timestampp: time_iso8601
       - labels:
           hostname:
       - timestamp:
           source: timestampp
           format: '2006-01-02T15:04:05-07:00'
```

**Comment by Mario-Hofstaetter** [comment]
@ufth just curious, what was the culprit? Besides the timestamp format?
I just diffed the configs, was it the typo in `acess` of the `selector`? But then you should not have gotten any logs because nothing matched?

![grafik](https://user-images.githubusercontent.com/33002073/117874372-2d096000-b2a1-11eb-8fab-8d6a8fd26282.png)

**Comment by ufth** [comment]
@Mario-Hofstaetter omg, I have never noticed this typo until you mentioned it now! Just spent a week of my life for nothing :-p

Got my old configuration file, fixed this typo and everything works smoothly with your format corrections. That's why "match" section was never actually applied by promtail.

> But then you should not have gotten any logs because nothing matched?

it seems that even without "match" section it just spits out any log line as it is to loki:

```
cat json_access.log.small | ./promtail --stdin -config.file promtail-local-config.yaml.99 -log.level=debug -dry-run
Clients configured:
----------------------
url: http://localhost:3100/loki/api/v1/push
batchwait: 1s
batchsize: 1048576
backoff_config:
  min_period: 500ms
  max_period: 5m0s
  max_retries: 10
timeout: 10s
tenant_id: ""

level=debug ts=2021-05-11T19:58:55.589370761Z caller=manager.go:58 msg="configured to read from stdin"
level=info ts=2021-05-11T19:58:55.589816651Z caller=server.go:229 http=[::]:9080 grpc=[::]:38271 msg="server listening on addresses"
level=info ts=2021-05-11T19:58:55.590228334Z caller=main.go:112 msg="Starting Promtail" version="(version=2.2.1, branch=HEAD, revision=babea82e)"
2021-05-11T22:58:55     {__path__="/var/log/nginx/json_access.log", host="yourhost", job="nginx_access_log"}    {"msec": "1619676109.299", "connection": "95359955", "connection_requests": "1", "pid": "25755", "request_id": "cf0a6bf1e469860b94fa76f48fd0236d", "request_length": "457", "remote_addr": "93.158.161.42", "remote_user": "", "remote_port": "", "time_local": "29/Apr/2021:09:01:49 +0300", "time_iso8601": "2021-04-29T09:01:49+03:00", "request": "GET /ru/yandex HTTP/1.1", "request_uri": "/ru/yandex", "args": "", "status": "200", "body_bytes_sent": "54964", "bytes_sent": "55720", "http_referer": "", "http_user_agent": "Mozilla/5.0 (compatible; YandexBot/3.0; +http://yandex.com/bots)", "http_x_forwarded_for": "93.158.161.42", "http_host": "example.com", "server_name": "example.com", "request_time": "0.111", "upstream": "127.0.0.1:9001", "upstream_connect_time": "0.000", "upstream_header_time": "0.108", "upstream_response_time": "0.112", "upstream_response_length": "159471", "upstream_cache_status": "EXPIRED", "ssl_protocol": "", "ssl_cipher": "", "scheme": "http", "request_method": "GET", "server_protocol": "HTTP/1.1", "pipe": ".", "gzip_ratio": "2.90", "http_cf_ray": "64765160fbf27a7d-DME", "request_completion": "OK"}
2021-05-11T22:58:55     {__path__="/var/log/nginx/json_access.log", host="yourhost", job="nginx_access_log"}    {"msec": "1619676114.113", "connection": "95359900", "connection_requests": "1", "pid": "25755", "request_id": "56ec9ddb6a2fda08c2ea7283c8ddc504", "request_length": "677", "remote_addr": "66.249.75.164", "remote_user": "", "remote_port": "", "time_local": "29/Apr/2021:09:01:54 +0300", "time_iso8601": "2021-04-29T09:01:54+03:00", "request": "GET /someurl", "args": "Name=bill", "status": "200", "body_bytes_sent": "53794", "bytes_sent": "54028", "http_referer": "", "http_user_agent": "Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)", "http_x_forwarded_for": "66.249.75.164", "http_host": "example.com", "server_name": "example.com", "request_time": "22.727", "upstream": "127.0.0.1:9002", "upstream_connect_time": "0.000", "upstream_header_time": "22.623", "upstream_response_time": "22.727", "upstream_response_length": "647782", "upstream_cache_status": "", "ssl_protocol": "", "ssl_cipher": "", "scheme": "http", "request_method": "GET", "server_protocol": "HTTP/1.1", "pipe": ".", "gzip_ratio": "12.04", "http_cf_ray": "647650f2ee470389-ORD", "request_completion": "OK"}
```
with this config file:
```
cat promtail-local-config.yaml.99
server:
  http_listen_port: 9080
  grpc_listen_port: 0

positions:
  filename: /data/loki/positions.yaml

clients:
  - url: http://localhost:3100/loki/api/v1/push

scrape_configs:
 - job_name: system
   static_configs:
   - targets:
      - localhost
     labels:
      job: nginx_access_log
      host: yourhost
      __path__: /var/log/nginx/json_access.log

```

**Comment by ufth** [comment]
> Ouch 😬, well at least its working now 😄

indeed! And I'm really happy that I don't feel that kind of frustration I experienced frequently when trying to add a bit comples to ansible YAML playbooks.

> Why is `selector: '{job="nginx_access_log"}'` there in the first place?
> You only have on job in `scrape_configs` or is this a limited example for github? Then it makes sense.

This example is actually full and my config is definitely ugly as I have no real experience with Loki.  So the answer is "I don't really know - I pasted some example to be sure I have some skeleton". My current use case is to get one single log file with all nginx virtualhosts/servers to loki, so I can get some relevant log lines when metrics that are already gathered by telegraf and stored in influxdb are scary.

But in the future I plan to add mysql and php slowlogs too (I hope that extracting few important things from multiline logs won't be super challenging :smile: ).

If by any chance you have enough expertise and experience to show the right way to create clean and beautiful configs for each of these cases - I will be grateful:
1. config for just one single nginx log json_access.log in JSON with timestamp from logline
2. config for 3 different log types (nginx, php slow, mysql slow) with timestamp from logline (we may consider them JSON to simplify)

> I just looked into the docs thinking how to actually dismiss loglines not matching a criteria, it seems thats whats the `drop` stage is there for: https://grafana.com/docs/loki/latest/clients/promtail/stages/drop/ but it uses regex to drop, and a negative regex is not that easy sometimes.

yeah, `drop` stage seems to be interesting and useful
